File: problems/3sum/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    
    let result = []    
    nums = nums.sort((a,b) => a-b);    
    let i=0;    

    while(i<nums.length-2){
       if(i==0 || nums[i-1]!==nums[i]){
           let j=i+1;
           let k=nums.length-1; 

            while(j<k){
                let first = nums[i], sec = nums[j], last = nums[k];
                let sum = first + sec + last;
                if(sum>0){
                    k--
                } else if(sum<0){
                    j++
                } else if(sum==0){
                    result.push([first, sec, last])
                    while(j<k && nums[j]==nums[j+1]) j++;
                    while(j<k && nums[k]==nums[k-1]) k--;
                    j++;
                    k--
                }
            }
       }
        i++        
    }
    return result
};

File: problems/3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n=len(nums)
        ans=[]
        for i in range(0, n):            
            if i!=0 and nums[i]==nums[i-1]:
                continue
            else:
                j=i+1
                k=n-1            
                while j<k:
                    val=nums[i]+nums[j]+nums[k]
                    if val==0:
                        ans.append([nums[i],nums[j],nums[k]])
                        while j<k and nums[j]==nums[j+1]:
                            j+=1
                        while j<k and nums[k]==nums[k-1]:
                            k-=1
                        j+=1
                        k-=1
                    elif val>0:
                        k-=1
                    else:
                        j+=1
        
        return ans
                